services/frontend/pkg/controller/validator.go

```diff
@@ -0,0 +1,114 @@
+	}
+
+	log.Info("Validating InputManifest")
+	if err := validateInputManifest(inputManifest); err != nil {
+		log.Error(err, "error validating InputManifest")
+		return err
+	}
+
+	return nil
+}
+
+// ValidateCreate defines the logic when a kubernetes obj resource is created
+func (v *ImputManifestValdator) ValidateCreate(ctx context.Context, obj runtime.Object) error {
+	return v.validate(ctx, obj)
+}
+
+// ValidateUpdate defines the logic when a kubernetes obj resource is updated
+func (v *ImputManifestValdator) ValidateUpdate(ctx context.Context, oldObj, newObj runtime.Object) error {
+	return v.validate(ctx, newObj)
+}
+
+// ValidateDelete defines the logic when a kubernetes obj resource is deleted
+func (v *ImputManifestValdator) ValidateDelete(ctx context.Context, obj runtime.Object) error {
+	return v.validate(ctx, obj)
+}
+
+// validateInputManifest takes v1beta.InputManifest, validate its structure
+// and returns an error when the validation will fail.
+// It doesn't validate .spec.Providers field.
+func validateInputManifest(im *v1beta.InputManifest) error {
+
+	var rawManifest manifest.Manifest
+
+	// Fill providers only with names, to check if there are defined
+	for _, p := range im.Spec.Providers {
+		switch p.ProviderType {
+		case v1beta.GCP:
+			rawManifest.Providers.GCP = append(rawManifest.Providers.GCP, manifest.GCP{Name: p.ProviderName})
+		case v1beta.AWS:
+			rawManifest.Providers.AWS = append(rawManifest.Providers.AWS, manifest.AWS{Name: p.ProviderName})
+		case v1beta.HETZNER:
+			rawManifest.Providers.Hetzner = append(rawManifest.Providers.Hetzner, manifest.Hetzner{Name: p.ProviderName})
+		case v1beta.OCI:
+			rawManifest.Providers.OCI = append(rawManifest.Providers.OCI, manifest.OCI{Name: p.ProviderName})
+		case v1beta.AZURE:
+			rawManifest.Providers.Azure = append(rawManifest.Providers.Azure, manifest.Azure{Name: p.ProviderName})
+		case v1beta.CLOUDFLARE:
+			rawManifest.Providers.Cloudflare = append(rawManifest.Providers.Cloudflare, manifest.Cloudflare{Name: p.ProviderName})
+		case v1beta.HETZNER_DNS:
+			rawManifest.Providers.HetznerDNS = append(rawManifest.Providers.HetznerDNS, manifest.HetznerDNS{Name: p.ProviderName})
+		}
+	}
+
+	rawManifest.Name = im.GetNamespacedName()
+	rawManifest.NodePools = im.Spec.NodePools
+	rawManifest.Kubernetes = im.Spec.Kubernetes
+	rawManifest.LoadBalancer = im.Spec.LoadBalancer
+
+	// Run the validation of all field except the Provider Fields.
+	// Providers will be validated separatly in the controller, after
+	// it will gather all the credentials from the K8s Secret resources
+	if err := rawManifest.Kubernetes.Validate(&rawManifest); err != nil {
+		return err
+	}
+	if err := rawManifest.NodePools.Validate(&rawManifest); err != nil {
+		return err
+	}
+	if err := rawManifest.LoadBalancer.Validate(&rawManifest); err != nil {
+		return err
+	}
+	if err := manifest.CheckLengthOfFutureDomain(&rawManifest); err != nil {
+		return err
+	}
+
+	return nil
```

Comment:
The provider validation is used later, I will try to address that somehow 👍

Reply:
Yes, I meant that the `Validate()` functions isn't used now, and it was used to validate the manifest, but since you've changed it that you validate the providers separately you can just change the `Validate()` function by removing this code block 
https://github.com/berops/claudie/blob/35d683e108a60171c9b6e80ca0e7051efa93e4e4/internal/manifest/validate.go#L15-L23
and just call `rawManifest.Validate()` here

Reply:
Will do, thanks for the clarification :)